src/core/thread/discover_scanner.hpp

```diff
@@ -0,0 +1,157 @@
+     * indicated by `aResult` pointer being set to NULL.
+     *
+     */
+    typedef otHandleActiveScanResult Handler;
+
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in]  aInstance     A reference to the OpenThread instance.
+     *
+     */
+    explicit DiscoverScanner(Instance &aInstance);
+
+    /**
+     * This method starts a Thread Discovery Scan.
+     *
+     * @param[in]  aScanChannels      Channel mask listing channels to scan (if empty, use all supported channels).
+     * @param[in]  aPanId             The PAN ID filter (set to Broadcast PAN to disable filter).
+     * @param[in]  aJoiner            Value of the Joiner Flag in the Discovery Request TLV.
+     * @param[in]  aEnableFiltering   Enable filtering MLE Discovery Responses that don't match our factory EUI64.
+     * @param[in]  aHandler           A pointer to a function that is called on receiving an MLE Discovery Response.
+     * @param[in]  aContext           A pointer to arbitrary context information.
+     *
+     * @retval OT_ERROR_NONE       Successfully started a Thread Discovery Scan.
+     * @retval OT_ERROR_NO_BUFS    Could not allocate message for Discovery Request.
+     * @retval OT_ERROR_BUSY       Thread Discovery Scan is already in progress.
+     *
+     */
+    otError Discover(const Mac::ChannelMask &aScanChannels,
+                     Mac::PanId              aPanId,
+                     bool                    aJoiner,
+                     bool                    aEnableFiltering,
+                     Handler                 aHandler,
+                     void *                  aContext);
+
+    /**
+     * This method indicates whether or not an MLE Thread Discovery Scan is currently in progress.
+     *
+     * @returns true if an MLE Thread Discovery Scan is in progress, false otherwise.
+     *
+     */
+    bool IsInProgress(void) const { return (mState != kStateIdle); }
+
+private:
+    enum State
+    {
+        kStateIdle,
+        kStateScanning,
+        kStateScanDone,
+    };
+
+    // Methods used by `MeshForwarder`
+    otError PrepareDiscoveryRequestFrame(Mac::TxFrame &aFrame);
+    void    HandleDiscoveryRequestFrameTxDone(Message &aMessage);
+    void    Stop(void) { HandleDiscoverComplete(); }
+
+    // Methods used from `Mle`
+    void HandleDiscoveryResponse(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo) const;
+
+    void        HandleDiscoverComplete(void);
+    static void HandleTimer(Timer &aTimer);
+    void        HandleTimer(void);
+
+    Handler                               mHandler;
+    void *                                mHandlerContext;
+    TimerMilli                            mTimer;
+    MeshCoP::SteeringData::HashBitIndexes mFilterIndexes;
+    Mac::ChannelMask                      mScanChannels;
+    uint8_t                               mScanChannel;
+    State                                 mState;
```

Comment:
nit: to avoid padding? Also, with C++11, we can specify a type for enums.
```suggestion
    State                                 mState;
    uint8_t                               mScanChannel;
```